crypto/keys/internal/ecdsa/privkey.go

```diff
@@ -7,9 +7,48 @@ import (
 	"crypto/sha256"
 	"fmt"
 	"math/big"
+
+	"golang.org/x/crypto/cryptobyte"
+	"golang.org/x/crypto/cryptobyte/asn1"
 )
 
-// GenPrivKey generates a new secp256r1 private key. It uses operating system randomness.
+// returns the curve order for the secp256r1 curve
+// NOTE: this is specific to the secp256r1/P256 curve,
+// and not taken from the domain params for the key itself
+// (which would be a more generic approach for all EC)
+
+func p256Order() *big.Int {
+	return elliptic.P256().Params().N
+}
+
+// returns half the curve order to restrict
+// signatures to low s
+
+func p256OrderDiv2() *big.Int {
+	return new(big.Int).Div(p256Order(), new(big.Int).SetUint64(2))
+}
```

Comment:
Reviewing some other code that implemented this normalization I saw that perhaps this code was a bit unidiomatic in that you can allocate these as global variables:

```golang
var p256Order = elliptic.P256().Params().N
var p256OrderHalf = new(big.Int).Rsh(p256Order, 1)
```

However, the only relevant parameter here is the curve order, so potentially it might make sense to retain functions for these operations which are expected to receive the curve order as a parameter.

Reply:
Yes, I saw that too. And I also agree with:

> However, the only relevant parameter here is the curve order, so potentially it might make sense to retain functions for these operations which are expected to receive the curve order as a parameter.

I think that this code should actually ask the key to give us its curve order (and then derive 1/2 from that). I wasn't sure that was possible with the interface I had, but I'll look into that now.

Reply:
Perhaps something like:

```golang
// Maximum scalar element in the "low half" of the field (inclusive)
func maxLowScalar(curveOrder *big.Int) *big.Int {
    new(big.Int).Rsh(curveOrder, 1)
}
```

Reply:
I thought it might be nice to use [`CurveParams`](https://pkg.go.dev/crypto/elliptic#CurveParams) as the parameter as and taking `params.N` as that would be rather hard to screw up, but I don't think the Go standard library defines the params for secp256k1.

Reply:
> // Maximum scalar element in the "low half" of the field (inclusive)
func maxLowScalar(curveOrder *big.Int) *big.Int {
    new(big.Int).Rsh(curveOrder, 1)
}

Shifting right 1 is an efficient way of dividing by 2. But wouldn't you hope that Div would already do the right thing? ;)

Reply:
It's six of one, half dozen of another, or as I said earlier, "a fancy function for dividing a `big.Int` in half"

Reply:
Perhaps not, but the compiler almost certainly does. I would prefer to leave this code as-is because it is:

a) Clearer to many (most?) programmers, than using Rsh
b) It is specifically choosing division. An Rsh is a bit shift, not a division, and thus not as clear an expression of what I want to do, to the compiler. 
c) Unless this division by 2 is a clear performance bottleneck, I would recommend it stays as-is.

Reply:
FWIW, I decided that rather than assuming one thing or another, I would test Div vs Rsh, so I wrote a basic program:

`package main

import (
	"crypto/elliptic"
	"math/big"
	"fmt"
	"time"
)


var p256Order = elliptic.P256().Params().N

func p256HalfOrder() (*big.Int) {
	return new(big.Int).Div(p256Order, new(big.Int).SetUint64(2))
}

func p256RshHalf() (*big.Int) {
	return new(big.Int).Rsh(p256Order, 1)
}

func main() {
	start1 := time.Now()
	fmt.Println("Div: ", p256HalfOrder())
	duration1 := time.Since(start1)
	start2 := time.Now()
	fmt.Println("Rsh: ", p256RshHalf())
	duration2 := time.Since(start2)
	fmt.Printf("Div: %v, Rsh: %v", duration1, duration2)
}
`

Rsh is roughly 10x the speed of Div, so I would say that there is no optimization applied by the compiler for Div with powers of 2, at least on GOOS=linux GOARCH=arm64.

Will make this change.

Reply:
yes, there is a big difference (in a proper benchmark, I suspect there can be even more difference, as DIV will do lot of unnecessary operations). Compiler can't magically guess that because we are using a complex structure, not a CPU word.
BTW: the best way to do a benchmark in Go is to use testing.Benchmark